src/main/java/com/commercetools/sync/commons/utils/QueryAll.java

```diff
@@ -1,166 +1,180 @@
-        return queryPage(client, 0).thenCompose(result -> {
-            final List<CompletableFuture<Void>> futureResults = new ArrayList<>();
-            consumer.accept(result.getResults());
-            futureResults.addAll(queryNextPages(client, result.getTotal(), consumer));
-            return CompletableFuture.allOf(futureResults.toArray(new CompletableFuture[futureResults.size()]));
-        });
+    CompletionStage<Void> run(@Nonnull final Consumer<List<T>> pageConsumer) {
+        this.pageConsumer = pageConsumer;
+        final CompletionStage<PagedQueryResult<T>> firstPage = queryPage();
+        return queryNextPages(firstPage).thenAccept(result -> { });
     }
 
     /**
-     * Given a {@link Consumer}, this method first calculates the total numbr of pages resulting from the query, then
-     * it applies this consumer on each page of the results from {@link this} instance's {@code baseQuery}
-     * and returns a list of futures after applying of this consumer on each page.
+     * Given a future containing a current page result {@link PagedQueryResult}, this method checks if the future is not
+     * null. If it is not, then it processes the result and attempts to fetch a next page, then it recursivley calls
+     * itself on the next page. If there is no next page, then the future would be null and this method would just
+     * return a future containing this instance of {@link QueryAll} as a result.
      *
-     * @param client        the CTP client that the query is run on.
-     * @param totalElements the total number of elements resulting from the query.
-     * @param consumer      the consumer to apply on each page of results.
-     * @return a list of futures of applying the consumer on each page of results.
+     * @param currentPageStage a future containing a result {@link PagedQueryResult}.
+     * @return a future containing this instance of {@link QueryAll} as a result.
      */
     @Nonnull
-    private List<CompletableFuture<Void>> queryNextPages(final SphereClient client, final long totalElements,
-                                                         final Consumer<List<T>> consumer) {
-        final long totalPages = getTotalNumberOfPages(totalElements);
-        return LongStream.range(1, totalPages)
-                         .mapToObj(page -> queryPage(client, page)
-                             .thenApply(PagedQueryResult::getResults)
-                             .thenAccept(consumer)
-                             .toCompletableFuture())
-                         .collect(toList());
+    private CompletionStage<QueryAll<T, C, S>> queryNextPages(
+        @Nullable final CompletionStage<PagedQueryResult<T>> currentPageStage) {
+        if (currentPageStage != null) {
+            return currentPageStage
+                .thenCompose(currentPage -> queryNextPages(processPageAndGetNext(currentPage)));
+        }
+        return completedFuture(this);
```

Comment:
maybe for this private member method it is a bit obfuscated to return `QueryAll<T, C, S> this`, because actual result is immutable in the instance (**this**) thus a bit confusing how to use the stage result.
I'd rather make it:
```java
CompletionStage<List<S>> run(@Nonnull final Function<List<T>, S> pageMapper) {
// ...
    return queryNextPages(firstPage)
        .thenApply(nextPage -> this.mappedResultsTillNow);
}

private CompletionStage<Void> queryNextPages(@Nullable final CompletionStage<PagedQueryResult<T>> currentPageStage) {
    // ...
    return completedFuture(null);
}```

Reply:
2638c02